client/src/logic/impl/MapNavigatorImpl.ts

```diff
@@ -1,19 +1,92 @@
-import { NavigationNode } from "../../types/NavigationNode";
+import { MapNodeFilter } from "../../types/roomsearch/MapNodeFilter";
+import { NavigationDirections } from "../../types/navigation/NavigationDirections";
+import { NavigationNode } from "../../types/navigation/NavigationNode";
 import { MapNavigator } from "../interfaces/MapNavigator";
+import { Graph } from "../interfaces/Graph";
+import { NeighbourConnection } from "../../types/graph/NeighbourConnection";
+import { Constants } from "../../utils/Constants";
 
 export class MapNavigatorImpl implements MapNavigator {
-    
+    private graph: Graph;
+
+    constructor(graph: Graph) {
+        this.graph = graph;
+    }
+
     findShortestPath(startNodeId: number, endNodeFilter: MapNodeFilter): NavigationDirections {
         let path = this.findPathToNearestNode(startNodeId, endNodeFilter);
         let navNodes = this.convertPathToNavigationNodes(path);
         return this.splitNavigationPathIntoSteps(navNodes);
     }
 
-    private findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
-        throw new Error("Method not implemented.");
+    /*private*/ findPathToNearestNode(startNodeId: number, endNodeFilter: MapNodeFilter): MapNode[] {
+
+        if (typeof this.graph.getNode(startNodeId) === "undefined") {
+            throw new Error("Start location not valid.");
+        }
+
+        let path: MapNode[] = [];
+        let neighbours: NeighbourConnection[];
+
+        let parents = new Map<number, number>(); //childId, parentId
+        let bestDistances = new Map<number, number>(); //nodeId, distance
+
+        let SortedSet = require('sorted-set');
+        let unvisitedNodes = new SortedSet({
+            hash(item: any) {return item.nodeId},
+            compare(a: any, b: any) {return a.distance - b.distance;}
+        });
+
+        let currentDistance: number;
+        let currentNodeId: number;
+        let allVisited: boolean;
+
+        bestDistances.set(startNodeId, 0);
+        unvisitedNodes.add({nodeId: startNodeId, distance: 0});
+
+        while (true) {
+            currentDistance = Constants.infinity;
+            allVisited = true;
+
+            if (unvisitedNodes.length > 0) {
+                currentNodeId = unvisitedNodes.head().nodeId;
+                currentDistance = unvisitedNodes.head().distance;
+                allVisited = false;
+            }
+
+            if (endNodeFilter.satisfies(this.graph.getNode(currentNodeId))) {
+                path.push(this.graph.getNode(currentNodeId));
+                
+                while (parents.has(currentNodeId)) {
+                    let parentId = parents.get(currentNodeId);
+
+                    path.push(this.graph.getNode(parentId));
+                    currentNodeId = parentId;
+                }
+
+                return path.reverse();
+
+            } else if (allVisited) {
+                throw new Error("Destination cannot be reached.");
+            }
```

Comment:
This is the end condition. It returns the result, but it is a bit confusing because it is written in the middle of the 
algorithm code. This code should be moved below the while.

You probably added this code for early stopping before all nodes are visited. You can still achieve that by calling `break` if the filter is satisfied.

Bonus tip - the path recreation algorithm (this little while-loop) is not an essential part of dijkstra and should be extracted into a function.